caravel/assets/javascripts/dashboard.jsx

```diff
@@ -182,15 +167,148 @@ class GridLayout extends React.Component {
+
+    this.addSlices = this.addSlices.bind(this);
+    this.toggleSlice = this.toggleSlice.bind(this);
+    this.slicesLoaded = false;
+  }
+
+  addSlices() {
+    var slices = this.state.slices.filter(function (slice) {
+      return slice.isSelected;
+    });
+    var sliceIds = [];
 
+    var sliceObjects = slices.map(function (slice) {
+      var sliceObj = px.Slice(slice.data, this.props.dashboard);
+      $("#slice_" + slice.data.slice_id).find('a.refresh').click(function () {
+        sliceObj.render(true);
+      });
+      this.props.dashboard.slices.push(sliceObj);
+      sliceIds.push(sliceObj.data.slice_id)
+      return sliceObj;
+    }, this);
+
+    this.props.dashboard.addSlicesToDashboard(sliceIds);
+  }
+
+  toggleSlice(sliceIndex) {
+    this.setState({
+      slices: update(this.state.slices, {
+        [sliceIndex]: {
+          isSelected: {
+            $set: !this.state.slices[sliceIndex].isSelected
+          }
+        }
+      })
+    });
+  }
+
+  componentDidMount() {
+    var uri = "/sliceaddview/api/read?_flt_0_created_by=" + this.props.dashboard.curUserId;
+    this.slicesRequest = $.get(uri, function (response) {
+      this.slicesLoaded = true;
+      this.setState({
+        slices: response.result
+      }, function () {
+        $("#add-slice-container table").DataTable();
+      });
+    }.bind(this));
+  }
+
+  componentWillUnmount() {
+    this.slicesRequest.abort();
+  }
+
+  render() {
+    return (
+      <div className="modal fade" id="add_slice_modal" role="dialog">
+          <div className="modal-dialog" role="document">
+              <div className="modal-content">
+                  <div className="modal-header">
+                      <button type="button" className="close" data-dismiss="modal" aria-label="Close">
+                        <span aria-hidden="true">&times;</span>
+                      </button>
+                      <h4 className="modal-title" id="myModalLabel">Add New Slices</h4>
+                  </div>
+                  <div className="modal-body">
+                    <img src={"/static/assets/images/loading.gif"}
+                         className={"loading " + (this.slicesLoaded ? "hidden" : "")}
+                         alt="loading"/>
+                    <table className="dataframe table table-condensed table-hover dataTable no-footer">
+                      <thead className={this.slicesLoaded ? "" : "hidden"}>
+                        <tr>
+                          <th>Name</th>
+                          <th>Viz</th>
+                          <th>Creator</th>
+                        </tr>
+                      </thead>
+                      <tbody>
+                        {this.state.slices.map(function (slice, i) {
+                          if (this.props.slicesOnDashMap[String(slice.data.slice_id)]) {
+                            return null;
+                          }
+
+                          return (
+                            <tr key={i}
```

Comment:
does the slice have an id? it's generally a better practice to use a unique id rather than the array index for react keys.

https://github.com/airbnb/javascript/tree/master/react#props